README.md
# rekognition-node

This is a fork of [node-rekognition](https://github.com/oleurud/node-rekognition) by [oleurud](https://github.com/oleurud)

Use this if you need to ability to send AWS image bytes in addition to an S3 Bucket

## Installation

    npm install rekognition-node

## Use

### Instantiation

```javascript
const Rekognition = require('rekognition-node');

// Set your AWS credentials
const AWSParameters = {
  accessKeyId: 'XXX',
  secretAccessKey: 'XXX',
  region: 'XXX',
  bucket: 'XXX',
  ACL: 'XXX' // optional
};

const rekognition = new Rekognition(AWSParameters);
```

The ACL is optional and its possible values are: "private", "public-read", "public-read-write", "authenticated-read", "aws-exec-read", "bucket-owner-read", "bucket-owner-full-control" [More info](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property)

### Upload images to S3

Some methods from AWS Rekognition need one or more images uploaded to AWS S3 bucket

```javascript
/**
 * Upload image or images array to S3 bucket into specified folder
 *
 * @param {Array.<string>|string} imagePaths
 * @param {string} folder a folder name inside your AWS S3 bucket (it will be created if not exists)
 */
const s3Images = await rekognition.uploadToS3(imagePaths, folder)
```

### detectLabels

```javascript
/**
 * Detects instances of real-world labels within an image
 *
 * @param {Object|Buffer} image
 * @param {string} threshold (optional. Defaults 50)
 */
const imageLabels = await rekognition.detectLabels(image)
```

### detectFaces

```javascript
/**
 * Detects faces within an image
 *
 * @param {Object|Buffer} image
 */
const imageFaces = await rekognition.detectFaces(image)
```

### compareFaces

```javascript
/**
 * Compares a face in the source input image with each face detected in the target input image
 *
 * @param {Object|Buffer} sourceImage
 * @param {Object|Buffer} targetImage
 * @param {string} threshold (optional. Defaults 90)
 */
const faceMatches = await rekognition.compareFaces(sourceImage, targetImage, threshold)
```

### detectModerationLabels

```javascript
/**
 * Detects explicit or suggestive adult content in image
 *
 * @param {Object|Buffer} image
 * @param {number} threshold (optional. Defaults 50)
 */
const moderationLabels = await rekognition.detectModerationLabels(image, threshold)
```

### createCollection

```javascript
/**
 * Creates a collection
 *
 * @param {string} collectionId
 */
const collection = await rekognition.createCollection(collectionId)
```

### deleteCollection

```javascript
/**
 * Deletes a collection
 *
 * @param {string} collectionId
 */
const collection = await rekognition.deleteCollection(collectionId)
```

### indexFaces

```javascript
/**
 * Detects faces in the input image and adds them to the specified collection
 *
 * @param {string} collectionId
 * @param {Object} s3Image
 */
const facesIndexed = await rekognition.indexFaces(collectionId, s3Image)
```

### listFaces

```javascript
/**
 * List the metadata for faces indexed in the specified collection
 *
 * @param {string} collectionId
 */
const faces = await rekognition.listFaces(collectionId)
```

### searchFaces

```javascript
/**
 * Searches in the collection for matching faces of faceId
 *
 * @param {string} collectionId
 * @param {string} faceId
 * @param {number} threshold (optional. Defaults 90)
 */
const faceMatches = await rekognition.searchFacesByFaceId(collectionId, faceId, threshold)
```

### searchFacesByImage

```javascript
/**
 * First detects the largest face in the image (indexes it), and then searches the specified collection for matching faces.
 *
 * @param {string} collectionId
 * @param {Object} s3Image
 * @param {number} threshold (optional. Defaults 90)
 */
const faceMatches = await rekognition.searchFacesByImage(collectionId, s3Image, threshold)
```

## Test

First of all, you must create a _parameters.json_ file and set your AWS parameters. You have an example file _parametrs.json.example_

* cp parameters.json.example parameters.json
* vim parameters.json

Then:

* npm install
* npm test

## Changelog

Releases are documented in the [NEWS file](./NEWS.md)

## Requirements

node >= 7.10.0

## Contributing

**You are welcome contribute via pull requests.**

## More info about AWS Rekognition

http://docs.aws.amazon.com/rekognition/latest/dg/API_Operations.html
http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Rekognition.html


NEWS.md
# 0.4.1
- No yet

# 0.4.0
- Buffer images support
- Remove images folder from public package

# 0.3.0
- add ACL option to uploaded files

# 0.2.0
- add deleteCollection method


test/index.js
'use strict';

const Mocha = require('mocha');
const mocha = new Mocha({});

const debug = require('debug')('rekognition-node:test');

// Test suites
mocha.addFile(__dirname + '/integration/S3.js');
mocha.addFile(__dirname + '/integration/Rekognition.js');

mocha
  .run()
  .on('test', function(test) {})
  .on('test end', function(test) {})
  .on('pass', function(test) {})
  .on('fail', function(test, err) {
    if (process.env.NOTIFY == 1) {
      debug({
        title: test.title,
        message: test.file
      });
    }
  })
  .on('end', function() {
    process.exit();
  });


test/integration/S3.js
'use strict'

const expect = require('chai').expect
const S3 = require('../../AWS/S3')
const parameters = require('../../parameters')

const debug = require('debug')('node-rekognition:test:S3')

const imagePath = __dirname + '/../../images/run.jpg'
const s3 = new S3(parameters.AWS)

describe('S3', function () {
    it('not existing file should be ok', async function () {
        const result = await s3.exists('not-existing-file')

        expect(result).to.be.false
    })

    it('upload file should be ok', async function () {
        const s3Image = await s3.upload(imagePath, parameters.defaultFolder)

        expect(s3Image).to.have.property('ETag')
        expect(s3Image).to.have.property('Location')
        expect(s3Image).to.have.property('Key')
        expect(s3Image).to.have.property('Bucket')
        expect(s3Image.Bucket).to.be.equal(parameters.AWS.bucket)
    })

    it('upload multiple files should be ok and ordered', async function () {
        this.timeout(10000)

        const fileNames = [
            'run.jpg',
            'Mark_Zuckerberg.jpg',
            'Mark_Zuckerberg_and_wife2.jpg',
            'Bill_Gates.jpg',
            'volley_beach.jpg'
        ]

        const imagePaths = fileNames.map( fileName => __dirname + '/../../images/' + fileName)

        const s3Images = await s3.uploadMultiple(imagePaths, parameters.defaultFolder)

        // s3Image.Key = parameters.defaultFolder + new Date().getTime() + '-' + fileName
        const indexOfFileName = parameters.defaultFolder.length + (new Date().getTime() + '-').length

        expect(s3Images).to.be.an('array')
        s3Images.forEach( (s3Image, index) => {
            expect(s3Image).to.have.property('ETag')
            expect(s3Image).to.have.property('Location')
            expect(s3Image).to.have.property('Key')
            expect(s3Image).to.have.property('Bucket')
            expect(s3Image.Bucket).to.be.equal(parameters.AWS.bucket)
            expect(s3Image.Key.indexOf(fileNames[index])).to.be.equal(indexOfFileName)
        })
    })
})


test/integration/Rekognition.js
'use strict'

const expect = require('chai').expect
const Rekognition = require('../../AWS/rekognition')
const parameters = require('../../parameters')
const fs = require('fs');

const debug = require('debug')('node-rekognition:test:Rekognition')

const rekognition = new Rekognition(parameters.AWS)

let imageFile, sourceImageFile, targetImageFile, billGatesFile, moderationFile, imageBuffer, sourceImageBuffer, targetImageBuffer, billGatesBuffer, moderationBuffer
const collectionId = new Date().getTime() + 'test'

describe('Rekognition', function () {
    this.timeout(10000)

    it('uploadToS3 should response ok', async function () {
        const imagePaths = [
            __dirname + '/../../images/run.jpg',
            __dirname + '/../../images/Mark_Zuckerberg.jpg',
            __dirname + '/../../images/Mark_Zuckerberg_and_wife2.jpg',
            __dirname + '/../../images/Bill_Gates.jpg',
            __dirname + '/../../images/volley_beach.jpg'
        ]

        const s3Images = await rekognition.uploadToS3(imagePaths, parameters.defaultFolder)
        imageFile = s3Images[0]
        sourceImageFile = s3Images[1]
        targetImageFile = s3Images[2]
        billGatesFile = s3Images[3]
        moderationFile = s3Images[4]

        imageBuffer = fs.readFileSync(imagePaths[0])
        sourceImageBuffer = fs.readFileSync(imagePaths[1])
        targetImageBuffer = fs.readFileSync(imagePaths[2])
        billGatesBuffer = fs.readFileSync(imagePaths[3])
        moderationBuffer = fs.readFileSync(imagePaths[4])
    })

    describe('detect labels', async function () {
        context('with an s3 image', async function () {
            it('should response ok and have labels', async function () {
                const imageLabels = await rekognition.detectLabels(imageFile)

                expect(imageLabels).to.have.property('Labels')
                expect(imageLabels.Labels).to.be.an('array')
                imageLabels.Labels.forEach(function (element) {
                    expect(element).to.have.property('Name')
                    expect(element).to.have.property('Confidence')
                })
            })
        })

        context('with an image Buffer', async function () {
            it('should response ok and have labels', async function () {
                const imageLabels = await rekognition.detectLabels(imageBuffer)

                expect(imageLabels).to.have.property('Labels')
                expect(imageLabels.Labels).to.be.an('array')
                imageLabels.Labels.forEach(function (element) {
                    expect(element).to.have.property('Name')
                    expect(element).to.have.property('Confidence')
                })
            })
        })
    })

    describe('detect faces', async function () {
        context('with an s3 image', async function () {
            it('should response ok and have details', async function () {
                const imageFaces = await rekognition.detectFaces(imageFile)

                expect(imageFaces).to.have.property('FaceDetails')
                expect(imageFaces.FaceDetails).to.be.an('array')
                imageFaces.FaceDetails.forEach(function (element) {
                    expect(element).to.have.property('BoundingBox')
                    expect(element).to.have.property('Landmarks')
                    expect(element).to.have.property('Pose')
                    expect(element).to.have.property('Quality')
                    expect(element).to.have.property('Confidence')
                })
            })
        })

        context('with an image Buffer', async function () {
            it('should response ok and have details', async function () {
                const imageFaces = await rekognition.detectFaces(imageBuffer)

                expect(imageFaces).to.have.property('FaceDetails')
                expect(imageFaces.FaceDetails).to.be.an('array')
                imageFaces.FaceDetails.forEach(function (element) {
                    expect(element).to.have.property('BoundingBox')
                    expect(element).to.have.property('Landmarks')
                    expect(element).to.have.property('Pose')
                    expect(element).to.have.property('Quality')
                    expect(element).to.have.property('Confidence')
                })
            })
        })
    })

    describe('compare faces', async function () {
        context('with an s3 image', async function () {
            it('of Mark_Zuckerberg and Mark_Zuckerberg_and_wife2 should match and unmatch', async function () {
                const faceMatches = await rekognition.compareFaces(sourceImageFile, targetImageFile)

                expect(faceMatches).to.have.property('FaceMatches')
                expect(faceMatches).to.have.property('UnmatchedFaces')
                expect(faceMatches.FaceMatches).to.be.an('array')
                expect(faceMatches.UnmatchedFaces).to.be.an('array')
                expect(faceMatches.FaceMatches.length).to.be.equal(1)
                expect(faceMatches.UnmatchedFaces.length).to.be.equal(1)
                faceMatches.FaceMatches.forEach(function (element) {
                    expect(element).to.have.property('Similarity')
                    expect(element).to.have.property('Face')
                })
            })

            it('of Mark_Zuckerberg and Bill_Gates should unmatch', async function () {
                const faceMatches = await rekognition.compareFaces(sourceImageFile, billGatesFile)

                expect(faceMatches).to.have.property('FaceMatches')
                expect(faceMatches).to.have.property('UnmatchedFaces')
                expect(faceMatches.FaceMatches).to.be.an('array')
                expect(faceMatches.UnmatchedFaces).to.be.an('array')
                expect(faceMatches.FaceMatches.length).to.be.equal(0)
                expect(faceMatches.UnmatchedFaces.length).to.be.equal(1)
                faceMatches.FaceMatches.forEach(function (element) {
                    expect(element).to.have.property('Similarity')
                    expect(element).to.have.property('Face')
                })
            })
        })

        context('with an image buffer', async function () {
            it('of Mark_Zuckerberg and Mark_Zuckerberg_and_wife2 should match and unmatch', async function () {
                const faceMatches = await rekognition.compareFaces(sourceImageBuffer, targetImageBuffer)

                expect(faceMatches).to.have.property('FaceMatches')
                expect(faceMatches).to.have.property('UnmatchedFaces')
                expect(faceMatches.FaceMatches).to.be.an('array')
                expect(faceMatches.UnmatchedFaces).to.be.an('array')
                expect(faceMatches.FaceMatches.length).to.be.equal(1)
                expect(faceMatches.UnmatchedFaces.length).to.be.equal(1)
                faceMatches.FaceMatches.forEach(function (element) {
                    expect(element).to.have.property('Similarity')
                    expect(element).to.have.property('Face')
                })
            })

            it('of Mark_Zuckerberg and Bill_Gates should unmatch', async function () {
                const faceMatches = await rekognition.compareFaces(sourceImageBuffer, billGatesBuffer)

                expect(faceMatches).to.have.property('FaceMatches')
                expect(faceMatches).to.have.property('UnmatchedFaces')
                expect(faceMatches.FaceMatches).to.be.an('array')
                expect(faceMatches.UnmatchedFaces).to.be.an('array')
                expect(faceMatches.FaceMatches.length).to.be.equal(0)
                expect(faceMatches.UnmatchedFaces.length).to.be.equal(1)
                faceMatches.FaceMatches.forEach(function (element) {
                    expect(element).to.have.property('Similarity')
                    expect(element).to.have.property('Face')
                })
            })
        })
    })

    describe('detect moderation labels', async function () {
        context('with an s3 image', async function () {
            it(' should response ok and have details', async function () {
                const moderationLabels = await rekognition.detectModerationLabels(moderationFile)

                expect(moderationLabels).to.have.property('ModerationLabels')
                expect(moderationLabels.ModerationLabels).to.be.an('array')
                moderationLabels.ModerationLabels.forEach(function (element) {
                    expect(element).to.have.property('Name')
                    expect(element).to.have.property('Confidence')
                    expect(element).to.have.property('ParentName')
                })
            })
        })

        context('with an image buffer', async function () {
            it(' should response ok and have details', async function () {
                const moderationLabels = await rekognition.detectModerationLabels(moderationBuffer)

                expect(moderationLabels).to.have.property('ModerationLabels')
                expect(moderationLabels.ModerationLabels).to.be.an('array')
                moderationLabels.ModerationLabels.forEach(function (element) {
                    expect(element).to.have.property('Name')
                    expect(element).to.have.property('Confidence')
                    expect(element).to.have.property('ParentName')
                })
            })
        })

    })



    it('create collection should response ok', async function () {
        const collection = await rekognition.createCollection(collectionId)

        expect(collection).to.have.property('CollectionArn')
        expect(collection).to.have.property('StatusCode')
        expect(collection.StatusCode).to.be.equal(200)
    })

    it('index faces of Mark_Zuckerberg_and_wife2 should response ok', async function () {
        const facesIndexed = await rekognition.indexFaces(collectionId, targetImageFile)

        expect(facesIndexed).to.have.property('FaceRecords')
        expect(facesIndexed.FaceRecords.length).to.be.equal(2)
        facesIndexed.FaceRecords.forEach(function (face) {
            expect(face).to.have.property('Face')
            expect(face).to.have.property('FaceDetail')
        })
    })

    it('search faces by faceId should response ok', async function () {
        const facesIndexed = await rekognition.indexFaces(collectionId, sourceImageFile)
        const newFaceId = facesIndexed.FaceRecords[0].Face.FaceId

        const faceMatches = await rekognition.searchFacesByFaceId(collectionId, newFaceId)

        expect(faceMatches).to.have.property('FaceMatches')
        expect(faceMatches.FaceMatches.length).to.be.equal(1)
        faceMatches.FaceMatches.forEach(function (face) {
            expect(face).to.have.property('Face')
            expect(face).to.have.property('Similarity')
        })
    })

    it('search faces by image should response ok', async function () {
        const faceMatches = await rekognition.searchFacesByImage(collectionId, sourceImageFile)

        expect(faceMatches).to.have.property('FaceMatches')
        expect(faceMatches.FaceMatches.length).to.be.equal(2)
        faceMatches.FaceMatches.forEach(function (face) {
            expect(face).to.have.property('Face')
            expect(face).to.have.property('Similarity')
        })
    })

    it('list faces should response ok', async function () {
        const faces = await rekognition.listFaces(collectionId)

        expect(faces).to.have.property('Faces')
        expect(faces.Faces.length).to.be.equal(3)
        faces.Faces.forEach(function (face) {
            expect(face).to.have.property('FaceId')
            expect(face).to.have.property('ImageId')
        })
    })

    it('delete collection should response ok', async function () {
        const collection = await rekognition.deleteCollection(collectionId)

        expect(collection).to.have.property('StatusCode')
        expect(collection.StatusCode).to.be.equal(200)
    })
})


AWS/S3.js
'use strict';

const AWS = require('aws-sdk');
const fs = require('fs');
const debug = require('debug')('rekognition-node:S3');

const VALID_FORMATS = ['jpg', 'jpeg', 'png'];

module.exports = class S3 {
  constructor(AWSParameters) {
    this.s3 = new AWS.S3({
      accessKeyId: AWSParameters.accessKeyId,
      secretAccessKey: AWSParameters.secretAccessKey,
      region: AWSParameters.region,
      params: {
        Bucket: AWSParameters.bucket
      }
    });

    this.ACL = AWSParameters.ACL;
  }

  /**
   * Upload file to S3 bucket into specified folder
   *
   * @param {string} filePath
   * @param {string} folder
   */
  async upload(filePath, folder) {
    const s3FilePath = this.getS3FullPath(filePath, folder);

    if (!await this.exists(s3FilePath)) {
      return new Promise((resolve, reject) => {
        const fileStream = fs.createReadStream(filePath);

        const params = {
          Key: s3FilePath,
          Body: fileStream
        };

        if (this.ACL) {
          params.ACL = this.ACL;
        }

        const opts = {
          queueSize: 10, // upload parts in parallel
          partSize: 1024 * 1024 * 10 //10Mb
        };

        this.s3
          .upload(params, opts)
          .on('httpUploadProgress', function(evt) {
            debug('upload-part', evt.loaded, '/', evt.total);
          })
          .send(function(err, data) {
            if (err) {
              debug('upload-error', err);
              reject(err);
            }

            debug('upload-finish', s3FilePath);
            resolve(data);
          });
      });
    } else {
      debug('file-already-exists');
      return {
        Bucket: this.AWSParameters.bucket,
        Key: s3FilePath
      };
    }
  }

  /**
   * Upload multiple files in parallel to S3 bucket into specified folder
   *
   * @param {Array.<strings>} filePaths
   * @param {string} folder
   * @returns {Promise.<Array>} of S3 files in the same input order
   */
  async uploadMultiple(filePaths, folder) {
    return await Promise.all(
      filePaths.map(filePath => this.upload(filePath, folder))
    );
  }

  /**
   * Check if file exists in S3
   *
   * @param {string} s3FilePath
   * @return {Promise.<boolean>}}
   */
  async exists(s3FilePath) {
    return new Promise((resolve, reject) => {
      this.s3.headObject({ Key: s3FilePath }, function(err, data) {
        if (!err) resolve(true);

        resolve(false);
      });
    });
  }

  /**
   * Creates the S3 path and checks the file format
   *
   * @param {string} filePath
   * @param {string} folder
   */
  getS3FullPath(filePath, folder = 'rekognition-node-folder/') {
    const fileName = new Date().getTime() + '-' + filePath.split('/').pop();
    const extension = fileName
      .split('.')
      .pop()
      .toLowerCase();

    if (VALID_FORMATS.indexOf(extension) !== -1) return folder + fileName;
    else throw new Error('Invalid file format');
  }
};


AWS/rekognition.js
'use strict'

const AWS = require('aws-sdk')
const debug = require('debug')('node-rekognition:Rekognition')
const S3 = require('./S3')

module.exports = class Rekognition {
    constructor(AWSParameters) {
        this.rekognition = new AWS.Rekognition({
            accessKeyId: AWSParameters.accessKeyId,
            secretAccessKey: AWSParameters.secretAccessKey,
            region: AWSParameters.region
        })

        this.s3 = new S3(AWSParameters)
        this.bucket = AWSParameters.bucket
    }

    /**
     * Upload image or images array to S3 bucket into specified folder
     *
     * @param {Array.<string>|string} imagePaths
     * @param {string} folder a folder name inside your AWS S3 bucket (it will be created if not exists)
     */
    async uploadToS3(imagePaths, folder) {
        if (Array.isArray(imagePaths))
            return await this.s3.uploadMultiple(imagePaths, folder)
        else
            return await this.s3.upload(imagePaths, folder)
    }

    /**
     * Do the request to AWS Rekognition
     *
     * @param {string} endpoint
     * @param {Object} params
     */
    doCall(endpoint, params) {
        return new Promise((resolve, reject) => {
            this.rekognition[endpoint](params, function (err, data) {
                if (err) {
                    reject(err)
                }
                else
                    resolve(data)
            })
        })
    }

    /**
     * Utility to get image params for s3 object or Bytes
     *
     * @param {Object|Buffer} image
     * @return {Object} image params for Rekognition
     */

    getImageParams(image) {
        return image instanceof Buffer
            ? {
                Bytes: image
            }
            : {
                S3Object: {
                    Bucket: this.bucket,
                    Name: image.Key
                }
            };
    }

    /**
     * Detects instances of real-world labels within an image
     *
     * @param {Object|Buffer} image
     * @param {string} threshold
    */
    async detectLabels(image, threshold = 50) {
        const params = {
            Image: this.getImageParams(image),
            MaxLabels: 4096,
            MinConfidence: threshold
        }

        return await this.doCall('detectLabels', params)
    }

    /**
     * Detects faces within an image
     *
     * @param {Object|Buffer} image
     */
    async detectFaces(image) {
        const params = {
            Image: this.getImageParams(image)
        }

        return await this.doCall('detectFaces', params)
    }

    /**
     * Compares a face in the source input image with each face detected in the target input image
     *
     * @param {Object|Buffer} sourceImage
     * @param {Object|Buffer} targetImage
     * @param {string} threshold
     */
    async compareFaces(sourceImage, targetImage, threshold = 90) {
        const params = {
            SimilarityThreshold: threshold,
            SourceImage: this.getImageParams(sourceImage),
            TargetImage: this.getImageParams(targetImage)
        }

        return await this.doCall('compareFaces', params)
    }

    /**
     * Detects explicit or suggestive adult content in image
     *
     * @param {Object|Buffer} image
     * @param {number} threshold
     */
    async detectModerationLabels(image, threshold = 50) {
        const params = {
            Image: this.getImageParams(image),
            MinConfidence: threshold
        }

        return await this.doCall('detectModerationLabels', params)
    }

    /**
     * Creates a collection
     *
     * @param {string} collectionId
     */
    async createCollection(collectionId) {
        const params = {
            CollectionId: collectionId
        }

        return await this.doCall('createCollection', params)
    }

    /**
     * Deletes a collection
     *
     * @param {string} collectionId
     */
    async deleteCollection(collectionId) {
        const params = {
            CollectionId: collectionId
        }

        return await this.doCall('deleteCollection', params)
    }

    /**
     * Detects faces in the input image and adds them to the specified collection
     *
     * @param {string} collectionId
     * @param {Object|Buffer} image
     */
    async indexFaces(collectionId, image) {
        var params = {
            CollectionId: collectionId,
            Image: this.getImageParams(image)
        }

        return await this.doCall('indexFaces', params)
    }

    /**
     * List the metadata for faces indexed in the specified collection
     *
     * @param {string} collectionId
     */
    async listFaces(collectionId) {
        var params = {
            CollectionId: collectionId,
            MaxResults: 4096
        }

        return await this.doCall('listFaces', params)
    }

    /**
     * Searches in the collection for matching faces of faceId
     *
     * @param {string} collectionId
     * @param {string} faceId
     * @param {number} threshold
     */
    async searchFacesByFaceId(collectionId, faceId, threshold = 90) {
        var params = {
            CollectionId: collectionId,
            FaceId: faceId,
            FaceMatchThreshold: threshold,
            MaxFaces: 4096
        }

        return await this.doCall('searchFaces', params)
    }

    /**
     * First detects the largest face in the image (indexes it), and then searches the specified collection for matching faces.
     *
     * @param {string} collectionId
     * @param {Object} s3Image
     * @param {number} threshold
     */
    async searchFacesByImage(collectionId, image, threshold = 90) {
        var params = {
            CollectionId: collectionId,
            Image: this.getImageParams(image),
            FaceMatchThreshold: threshold,
            MaxFaces: 4096
        }

        return await this.doCall('searchFacesByImage', params)
    }
}
